internal/controller/thanosreceive_controller.go

```diff
@@ -18,45 +18,193 @@ package controller
-// the ThanosReceive object against the actual cluster state, and then
-// perform operations to make the cluster state reflect the state specified by
-// the user.
-//
 // For more details, check Reconcile and its Result here:
 // - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.17.3/pkg/reconcile
 func (r *ThanosReceiveReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
-	_ = log.FromContext(ctx)
+	logger := log.FromContext(ctx)
+
+	// Fetch the ThanosReceive instance to validate it is applied on the cluster.
+	receiver := &monitoringthanosiov1alpha1.ThanosReceive{}
+	err := r.Get(ctx, req.NamespacedName, receiver)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			logger.Info("thanos receive resource not found. ignoring since object may be deleted")
+			return ctrl.Result{}, nil
+		}
+		logger.Error(err, "failed to get ThanosReceive")
+		return ctrl.Result{}, err
+	}
+
+	isMarkedForDeletion := receiver.GetDeletionTimestamp() != nil
+	if isMarkedForDeletion {
+		return r.handleDeletionTimestamp(logger, receiver)
+	}
 
-	// TODO(user): your logic here
+	err = r.syncResources(ctx, *receiver)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
 
 	return ctrl.Result{}, nil
 }
 
+// +kubebuilder:rbac:groups=monitoring.thanos.io,resources=thanosreceives,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=monitoring.thanos.io,resources=thanosreceives/status,verbs=get;update;patch
+// +kubebuilder:rbac:groups=monitoring.thanos.io,resources=thanosreceives/finalizers,verbs=update
+// +kubebuilder:rbac:groups=apps,resources=statefulsets;deployments,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups="",resources=services;configmaps;serviceaccounts,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups="discovery.k8s.io",resources=endpointslices,verbs=get;list;watch
+
 // SetupWithManager sets up the controller with the Manager.
 func (r *ThanosReceiveReconciler) SetupWithManager(mgr ctrl.Manager) error {
-	return ctrl.NewControllerManagedBy(mgr).
+	bld := ctrl.NewControllerManagedBy(mgr)
+	return r.buildController(*bld)
+}
+
+// buildController sets up the controller with the Manager.
+func (r *ThanosReceiveReconciler) buildController(bld builder.Builder) error {
+	bld.
 		For(&monitoringthanosiov1alpha1.ThanosReceive{}).
-		Complete(r)
+		Owns(&corev1.ConfigMap{}).
+		Owns(&corev1.ServiceAccount{}).
+		Owns(&corev1.Service{}).
+		Owns(&appsv1.Deployment{}).
+		Owns(&appsv1.StatefulSet{})
+
+	return bld.Complete(r)
+}
+
+// syncResources syncs the resources for the ThanosReceive resource.
+// It creates or updates the resources for the hashrings and the router.
+func (r *ThanosReceiveReconciler) syncResources(ctx context.Context, receive monitoringthanosiov1alpha1.ThanosReceive) error {
+	var objs []client.Object
+	logger := log.FromContext(ctx)
+
+	objs = append(objs, r.buildHashrings(receive)...)
+
+	var errCount int32
```

Comment:
i think it might become unwieldy - we already log on each of the failed requests but we will log the count at the top level. if anything we could drop it entirely but i think its a useful summary of the reconciliation loop?